contracts/pair_stable/src/contract.rs

```diff
@@ -591,7 +626,82 @@ pub fn withdraw_liquidity(
     }
 
     let (pools, total_share) = pool_info(deps.querier, &config)?;
-    let refund_assets = get_share_in_assets(&pools, amount, total_share);
+
+    let refund_assets = if let Some(assets) = assets {
+        // Imbalanced withdraw
+        if assets
+            .iter()
+            .any(|asset| !config.pair_info.asset_infos.contains(&asset.info))
+        {
+            return Err(StdError::generic_err("Invalid assets").into());
+        };
+        if assets[0].info == assets[1].info {
+            return Err(ContractError::DoublingAssets {});
+        }
+
+        let amp = compute_current_amp(&config, &env)?;
+        let d0 = compute_d(amp, pools[0].amount.u128(), pools[1].amount.u128()).unwrap();
+
+        let mut new_balances = pools
+            .iter()
+            .map(|pool| {
+                let pool_asset = if pool.info == assets[0].info {
+                    &assets[0]
+                } else {
+                    &assets[1]
+                };
+
+                pool.amount
+                    .checked_sub(pool_asset.amount)
+                    .map_err(Into::into)
+            })
+            .collect::<StdResult<Vec<_>>>()?;
+        let d1 = compute_d(amp, new_balances[0].u128(), new_balances[1].u128()).unwrap();
+
+        // Get fee info from the factory
+        let fee_info = query_fee_info(
+            &deps.querier,
+            &config.factory_addr,
+            config.pair_info.pair_type.clone(),
+        )?;
+
+        let fee = fee_info.total_fee_rate / Uint128::from(2u8);
+
+        for i in 0..=1usize {
+            let ideal_amount = Uint128::from(d1).multiply_ratio(pools[i].amount, d0);
+
+            let difference = if ideal_amount > new_balances[i] {
+                ideal_amount - new_balances[i]
+            } else {
+                new_balances[i] - ideal_amount
+            };
+
+            let withdraw_fee = fee.checked_mul(difference)?;
+            new_balances[i] = new_balances[i].checked_sub(withdraw_fee)?;
+        }
+
+        let d2 = compute_d(amp, new_balances[0].u128(), new_balances[1].u128()).unwrap();
+
+        let need_lp_tokens = Uint128::from(d0)
+            .checked_sub(d2.into())?
+            .multiply_ratio(total_share, d0)
+            .checked_add(Uint128::from(1u8))?; // In case of rounding errors - make it unfavorable for the "attacker";
+        if need_lp_tokens > lp_amount_to_convert {
+            return Err(StdError::generic_err(format!(
+                "Not enough LP tokens. You need {} LP tokens.",
+                need_lp_tokens
+            ))
+            .into());
+        } else if need_lp_tokens.is_zero() {
+            return Err(
+                StdError::generic_err("Failed to calculate necessary LP tokens amount").into(),
+            );
+        }
```

Comment:
This block won't be ever executed due to line 688. The check should be done before adding one at line 688

Reply:
That's true. Thanks

Reply:
Need to fix it for pair_stable_bluna too